### PR TITLE
Kernel: First cut of a sb16 driver

### PR DESCRIPTION
**Pull request by rburchell**
Also add an AudioServer that (right now) doesn't do much.
It tries to open, parse, and play a wav file. In the future, it can do more.

My general thinking here here is that /dev/audio will be "owned" by AudioServer,
and we'll do mixing in software before passing buffers off to the kernel
to play, but we have to start somewhere.

### CODE REVIEW

File: Kernel/Devices/SB16.cpp [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+    dsp_write(0xd0);
+
+    IO::in8(DSP_STATUS); // 8 bit interrupt
+    if (m_vmaj >= 4)
+        IO::in8(DSP_R_ACK); // 16 bit interrupt
+}
+
+void SB16::initialize()
+{
+    int32_t i;
```
Comment (awesomekling) [inline Kernel/Devices/SB16.cpp]: This feels very C-like, can't we just scope this to the loop?

File: Kernel/Devices/SB16.cpp [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+}
+
+void SB16::dma_start(uint32_t length)
+{
+    const auto addr = m_dma_buffer_page->paddr().get();
+    const u8 channel = 1;
+    const u8 mode = 0;
+
+    // Disable the DMA channel
+    const int BIT_4 = 0x04;
```
Comment (awesomekling) [inline Kernel/Devices/SB16.cpp]: `BIT_4` seems like an odd name for the value `0x04`. Does it mean something in the context of the SB16?

File: Kernel/Devices/SB16.cpp [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+{
+    if (!m_dma_buffer_page) {
+        kprintf("SB16: Allocating page\n");
+        m_dma_buffer_page = MM.allocate_supervisor_physical_page();
+    }
+
+    kprintf("SB16: Writing buffer of %d bytes\n", length);
+    const int BLOCK_SIZE = 32 * 1024;
+    if (length > BLOCK_SIZE) {
+        return -ENOSPC;
```
Comment (awesomekling) [inline Kernel/Devices/SB16.cpp]: I wonder if there is a more appropriate error code for this. I always find it difficult to pick these.
Maybe `ENOSPC` is fine.

File: Kernel/Devices/SB16.cpp [hunk 4]
```diff
@@ -0,0 +1,183 @@
+#include "SB16.h"
+#include "IO.h"
+#include <Kernel/VM/MemoryManager.h>
+
+const int DSP_READ = 0x22A;
```
Comment (awesomekling) [inline Kernel/Devices/SB16.cpp]: I/O ports are 16-bit unsigned values, so let's just make them `u16`.

File: Kernel/Devices/SB16.cpp [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+
+    kprintf("SB16: Writing buffer of %d bytes\n", length);
+    const int BLOCK_SIZE = 32 * 1024;
+    if (length > BLOCK_SIZE) {
+        return -ENOSPC;
+    }
+    const u8 mode = 0;
+
+    disable_irq();
+    const int sample_rate = 44000;
```
Comment (awesomekling) [inline Kernel/Devices/SB16.cpp]: Should this be `44100`?

File: Kernel/Devices/SB16.h [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    void initialize();
+    void wait_for_irq();
+    void dma_start(uint32_t length);
+    void set_sample_rate(uint16_t hz);
+    void dsp_write(uint8_t value);
+    uint8_t dsp_read();
+
+    RefPtr<PhysicalPage> m_dma_buffer_page;
+    bool m_interrupted { false };
+    int m_vmaj { 0 };
```
Comment (awesomekling) [inline Kernel/Devices/SB16.h]: `m_major_version`?

File: Kernel/Devices/SB16.cpp (hunk 2 above)
Comment (rburchell) [reply Kernel/Devices/SB16.cpp]: https://homepages.cae.wisc.edu/~brodskye/sb16doc/sb16doc.html has been the best source I've been working from. This was a leftover from trial-and-error, I've fixed it accordingly.

File: Kernel/Devices/SB16.cpp (hunk 3 above)
Comment (rburchell) [reply Kernel/Devices/SB16.cpp]: ENOSPC feels right to me. You're trying to fit something too large into too small a space.

File: Kernel/Devices/SB16.cpp (hunk 5 above)
Comment (rburchell) [reply Kernel/Devices/SB16.cpp]: Kind of academic at the current state, but yes :)

**Pull request merged**